feat(protocol): support atlantic odyssee heatpumps

tspopp opened the pull request:
**This PR adds support for the protocol observed within an Atlantic Odyssee device #68.**

## Help

:warning: This PR needs help by the community, as I don't have a machine with the new protocol :warning: 

## Missing Features / Open Tasks:
- [x] Checksum Calculation
- [x] HMI Message
  - [x] time/date 
  - [x] target temperature
  - [x] operation types (timer/always-on) NEW: off-peak hours
  - [x] operation modes (boost, absence, eco...)
  - [x] air-duct config (int/int..)
  - [x] anti-legionella setting
  - [x] enable emergency mode
  - ~~disable heat-element~~
  - ~~enable pv-input~~ ~~
  - ~~timer window a/b~~
  - ~~test modes~~
  - ~~fan exhaust config (INT/EXT)~~
  - [x] error request id / error number 
- [x] Main Message
  - [x] hot water temp
  - [x] NEW: upper hot water temp
  - [x] NEW: lower hot water temp  
  - [x] input air temp
  - [x] lower evaporator temp
  - [x] upper evaporator temp
  - [x] compressor outlet temp
  - [x] states: heatpump, heatelement, ext, fan, defrost
  - ~~states: pv-input, solar-input...~~
  - [x] fan pwm 
- [x] Energy Message
  - [x] total heatpump hours
  - ~~total hours~~ 
  - [x] total hours heat element
  - ~~power heatpump~~
  - ~~power heatelement~~
  - [x] power total
  - ~~water production~~
  - [x] total energy
  - ~~diagnosis temperature values (min/max observed)~~
- [x] Error Message
  - [x] errorCode
  - [x] requestId 
  - [x] optional: error message content (documented, not implemented)
- [x] Extra Message 217
  - [x] power total
  - [x] NEW: grid voltage
  - [x] NEW: amperage 
- [x] Compatibility to legacy and next protocol
- [x] MITM
  - [x] Clarify origin of new message 217 
  - [x] Clarify frequences of emitted messages
  - [x] Replay message 217 in correct frequency 
  - [x] Test Passthrough
  - [x] Test Error Passthrough
  - [x] Test Override Values 
  - [x] PV Modes: If we can't disable the heat-element, we don't have "PV Mode HP Only". Determine alternative solutions...
- [x] MQTT Discovery depending on protocol version

## How to help?

This branch contains a modified version of AquaMQTT which is meant to be installed in LISTENER mode. It is currently able to identify hmi, main, energy and extra messages from the new protocol #68 . These are provided to mqtt on the three topics:

`aquamqtt/hmi/debug`
`aquamqtt/main/debug`
`aquamqtt/energy/debug`
`aquamqtt/extra/debug`
`aquamqtt/error/debug`

The new protocol findings are documented within [PROTOCOL_ODYSSEE.md](https://github.com/tspopp/AquaMQTT/blob/t/odyssee/PROTOCOL_ODYSSEE.md)

## Tracing

Using the debug python script https://github.com/tspopp/AquaMQTT/blob/main/tools/debug.py you are able to record changing messages over time and identify which location holds what kind of attribute.

### Create the python environment

```
python3 -m venv venv
source venv/bin/activate
pip install paho-mqtt
```

### Run the python script

```
source venv/bin/activate
python3 debug.py
```


## How to help here?

- start a recording of the serial messages
- change a single (previously unidentified) value at the hmi, e.g. water target temperature from 50 to 52
- stop the recording of the serial message
- attach the traces together with a description ("changed water target temperature from 50 to 52 and back")

FSlunkk commented:
Hi, i open the super secret engineering menu to find the error message but i don't find it. I put the debug file as an attachment. It's at the end of the file 7:39.
[aquadbg_debug_dec.csv](https://github.com/user-attachments/files/18746609/aquadbg_debug_dec.csv)
Thanks

FSlunkk commented:
I completed some data on the protocol table. 
However, I can't find the matches for the max and min values.

194 29 54 2 0 33 0 0 0 0 0 0 0 0 0 0 0 0 43 50 39 7 0 0 0 0 65 66 0 0 50 
67 25 65 0 1 16 0 0 0 0 196 0 0 0 182 46 0 0 0 0 0 0 0 0 0 0 21 
217 31 0 0 233 0 1 0 0 0 242 222 88 0 14 1 183 0 0 1 0 0 2 0 0 0 0 0 0 0 0 0 56 
193 29 19 0 14 0 12 0 11 0 13 0 0 0 36 0 14 0 0 0 0 0 0 2 0 0 0 0 82 66 50 

![PXL_20250211_063925155](https://github.com/user-attachments/assets/1d3bf6fd-0945-49d5-8bdf-1c7f14b0fac4)

I think that 217 is also main "233" seems to be the voltage and the "1" after the power in Watt. I will confirm this soon.

tspopp commented:
I've added the latest findings based on the document in your branch. Let me know if it works :)

FSlunkk commented:
This is what i found for the byte 2 : 

off-peak hours
Boost : 10 with byte n°14 duration in days
Eco off : 2
Eco on : 1
Auto : 0
Absence : 6 with byte n°13 duration in days

Continuous
Boost : 72 with byte n°14 duration in days
Eco off : 66
Eco on : 65
Auto : 64
Absence : 68 with byte n°13 duration in days

Timer
Boost : 24 with byte n°14 duration in days
Eco off : 18
Eco on : 17
Auto : 16
Absence : 20 with byte n°13 duration in days

When i actived the super secret menu :
194 29 54 2 0 33 0 0 0 0 0 0 0 0 0 0 0 0 43 50 15 18 0 0 0 0 65 66 15 46 46 
74 68 46 24 54 0 75 0 23 0 24 0 35 0 0 0 54 0 55 0 65 3 0 0 0 0 0 0 0 0 81 38 0 0 152 0 0 0 123 12 0 0 47 13 0 0 152 26 0 0 46 1 0 0 221 46 58 12 1 24 0 0 0 0 0 0 0 0 43 128 
67 25 65 0 6 16 0 0 0 0 196 0 0 0 187 46 0 0 0 0 0 0 0 0 0 0 31 
217 31 0 0 234 0 2 0 127 1 237 235 88 0 14 1 194 0 0 1 0 0 2 0 0 0 0 0 0 0 0 0 25 
193 29 53 0 37 0 18 0 18 0 23 0 0 0 53 0 46 0 0 0 0 0 0 2 0 0 0 0 82 66 19 

And i disconnected the HMI, the 217 line is still here.

FSlunkk commented:
Your mod works in listenner mode.

I found this : 

193 :
Byte 18 : Fan speed : 50 low / 100 high

217 : 217 31 209 0 232 0 230 1 210 0 77 241 88 0 14 1 208 0 0 1 0 0 2 0 0 0 0 0 0 0 0 0 43 
Byte 3 : tension in Volt 230 Volts
Bytes 5 et 6 : power en watts : 230 (Byte5) + 1 (byte6) *256 = 486 watts

194 :
Byte 4 : Anti-Legionella Mode / AirDuct Mode
Off and Int/Int : 0
On and Int/Int : 1
Off and Int/Ext : 16
On and Int/Ext : 17
Off and Ext/ext : 32
On and Ext/Ext : 33

Byte 5 : 
Emergency-Mode : Off : 0 / On : 1

194 29 54 2 0 33 0 0 0 0 0 0 0 0 0 0 0 0 44 50 52 16 0 0 0 0 65 66 0 0 49 

Byte 17 :
Day and month like your protocol
Byte 18
Year like yout protocol
Byte 20
Minutes
Byte 21
Hours

tspopp commented:
Great news. Just incorporate your findings in the document and I will take care of adding them to the implementation. According to your finding there are many similarities to the NEXT protocol :+1:

Regarding the super secret menu / error messages, I need you to go back to AquaDebug once and dump a whole series of messages (.csv) while opening the menu. I need to determine how the HMI is actually requesting errors. Maybe you can also share the amount of errors you have and create a photo of one of the errors shown in the super secret menu.

FSlunkk commented:
[super secret menu.csv](https://github.com/user-attachments/files/18773046/super.secret.menu.csv)
There is the whole serie of message. Is this fine for you?

tspopp commented:
These dumps are perfect :+1: I will add error message handling soon most likely tomorrow evening.

Do you have a [logic analyzer](https://www.az-delivery.de/en/products/saleae-logic-analyzer)? To replicate the serial behavior in MITM mode we should use the same timing for the messages. And since there is a new message, I am not sure if the current behavior is still correct. If you don't have one, I might add some code which measures the frequency, but if you have a logic analyzer it would be probably easier...

tspopp commented:
No worries, we will get there :+1:

tspopp commented:
>  byte 2 :
> 
> off-peak hours
> Boost : 10 with byte n°14 duration in days
> Eco off : 2
> Eco on : 1
> Auto : 0
> Absence : 6 with byte n°13 duration in days

Are you sure with the values for Boost and Absence? I'd expected something else :thinking: 

> Byte 4 : Anti-Legionella Mode

Is there an option to set the number of runs of the anti-legionella mode per month or do you just have switch on/off? 

> 217 :
> 
> 217 31 209 0 232 0 230 1 210 0 77 241 88 0 14 1 208 0 0 1 0 0 2 0 0 0 0 0 0 0 0 0 43
> Byte 3 : tension in Volt 230 Volts

You sure that this is really the voltage of the grid? We never had such thing before, it could be something else?

FSlunkk commented:
Yes, i'm sure for Boost and Absence.

For the anti-legionella mode, it's just off or 1 per month.

I'm pretty sure it's the voltage. The readings I took matched.

tspopp commented:
Ok, intresting that there are so many differences between these heatpumps. How old is that machine. Is it from 2024?

Actually I implemented and added most of the items from your last comments. Hope it works so far :+1:

FSlunkk commented:
My machine is from 2015.

tspopp commented:
:exploding_head:

FSlunkk commented:
It's old but it could be useful to other people who want to automate their old home.

tspopp commented:
Yeah, absolutely. I just thought this is next-gen. Instead this is last-gen, but you are right. Do these old heatpumps have the option to use cozytouch?

Regarding power consumption: Previous heatpumps had "totalPower" (heatpump and heatelement combined) and "powerHeatpump" and "powerHeatelement". Currently it is not clear if there is "totalPower" or "powerHeatpump" in Byte 5/6 of 217. You may want to run your machine on BOOST to identify it properly. 

Latest version should have all the things implemented you've provided, except power and grid voltage. Please test :)

FSlunkk commented:
![image](https://github.com/user-attachments/assets/7ec6b831-c6ec-4cbc-a873-1acb0b57bdc5)
I loaded the last version. It's strange because some data is missing, initially I only had water temp upper, then water temp, then compressor outlet temp.

FSlunkk commented:
![image](https://github.com/user-attachments/assets/3ba01fbc-2191-4d57-9197-fa2965b82385)
And now, i have this. Not all data is updated at the same time.

tspopp commented:
Ah, this looks very well :+1: Actually it is expected behavior: Updates to the MQTT broker are only send in case a value changes. Additionally there is a fixed time value set by `Configuration.h` which causes full update of all attributes:

```
/**
 * Change the time interval where all known attributes are re-published to the MQTT broker.
 */
constexpr uint32_t MQTT_FULL_UPDATE_MS = 1000 * 60 * 30;
```
If you want all attributes to be updated every minute, you can set this for exampel to `1000 * 60`. 

Since backends like homeassistant cache the last value, AquaMQTT does not need to rebroadcast values which haven't changed. 

Moreover, temperature values are filtered using kalman, so we remove the temperature sensor noise to reduce the amount of  updates even more.

FSlunkk commented:
Ok, good! I test boost tomorrow.

FSlunkk commented:
I tested Boost, Byte 5/6 of 217 is totalpower.

tspopp commented:
I added some code to determine the timing of the frames. I have not tested this on my heatpump, yet. But in theory, it should provide you frame timings e.g. to `aquamqtt/stats/timing/194-193` In this case, time it took after we received 194 and waited for message 193 in milliseconds. There should be multiple message timings such as `194-193`, `193-217` etc.

If this works, you may want to open the super secret menu once (so we also have the timing of the error message as well) and post a screenshot of MQTT explorer will all the timing values visible.

I think I can deduce the message frequency from this information, needed by the MITM mode. Other possibility is a logic analyzer :shrug:. But let's try this first, maybe we can get it work without that.

FSlunkk commented:
I loaded the last version but there is not aquamqtt/stats/timing in MQTT explorer. And i don't see Total Power and Grid voltage.
![image](https://github.com/user-attachments/assets/f5873c59-4433-4517-aebb-ba791dc3617e)

tspopp commented:
Are you sure flashing was ok and your branch is up to date, too? I also fixed time and date and it is also not visible in the screenshot :thinking:

FSlunkk commented:
I did it twice, I just did it again and it works.
![image](https://github.com/user-attachments/assets/9a00afdc-392f-4623-8715-fc23e247392e)
![image](https://github.com/user-attachments/assets/49d40f72-d4ff-4c1b-a1e7-93496c8ccdc5)
![image](https://github.com/user-attachments/assets/50d4bfb4-0bb1-4ff3-af89-0c1faef4f8ad)
![image](https://github.com/user-attachments/assets/b0caeb31-9440-4b83-a829-dd6cdbf2a94b)
![image](https://github.com/user-attachments/assets/09872319-3b0e-45d9-b1ad-a5a3bbef8ab9)
![image](https://github.com/user-attachments/assets/f2bb6ca9-556b-4e35-9a8f-1410cc27760c)

tspopp commented:
@FSlunkk If you have time to provide more protocol findings, I think we can get complete this issue soon. Also see missing items in the root description of this PR. Looking forward ;)

FSlunkk commented:
Yes, i can try to find more protocol but i think I'm going to need some help because I haven't found anything else obvious before.

tspopp commented:
Have you tried AquaMQTT (not AquaDebug) with serial debugging enabled and the Python script linked above?  It enables you to inspect the four messages independent from another, which makes it more easy to spot changing values. If you do traces while the heat pump is running you should be able to find the energy values by looking for upcounting values. Of course you can attach traces from a long run and I can see if I can locate something 🤷 

For MITM we should also identify the "disable heat element" option from the HMI, since that one is used by the PV Override Modes. I will prepare something to test MITM soon.

FSlunkk commented:
I try to collect debug message from aquamqtt/energy/debug but it's strange, I ran the piton collect file but it only worked once.
It seems like it can't retrieve the messages.
[aquamqtt_energy_debug_dec.csv](https://github.com/user-attachments/files/18949749/aquamqtt_energy_debug_dec.csv)

tspopp commented:
Actually you change the config to `client.subscribe('aquamqtt/+/debug')`. The `+` acts as wildcard and you will collect all the other messages (energy, extra, main, hmi), too. If these attributes we are looking for are not within the energy message, I suspect them to be in the extra message. Moreover, it does not collect a message if it hasn't changed. So only changing messages will appear in the logs. If it does not work, it may be due to not changed messages...

I will look in your traces later.

Regarding MITM: I looked already into it and it seems the protocol sequence is different. In the legacy and next protocol we have the message sequence `194 - 67 - 193 - 67`, while in yours it is `193 - 74-  57 - 217 - 193`. So I need to rewrite the MITM Tasks to apply to both to keep compatibility. I had very limited time for this project, but I try to implement it as soon there is free time.

FSlunkk commented:
2025-02-27 15:12:07.757161,25 67 0 79 16 0 0 0 0 198 0 0 0 4 47 0 0 0 0 0 0 0 0 0 0
So Byte 14-17 (4 47 0 0) is Total Operation Heat Pump 12036 hours.
![PXL_20250227_145528671](https://github.com/user-attachments/assets/d4791bae-2d06-4773-830d-9573e9014ceb)
and byte 10-13 (198 0 0 0) seem to be Total Operation Heatelement 198 hours.

FSlunkk commented:
2025-02-27 14:07:11.303284,25 67 0 78 16 0 0 0 0 198 0 0 0 2 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 14:12:08.219491,25 67 0 78 16 0 0 0 0 198 0 0 0 3 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 15:07:09.633552,25 67 0 79 16 0 0 0 0 198 0 0 0 3 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 15:12:07.757161,25 67 0 79 16 0 0 0 0 198 0 0 0 4 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 16:01:12.585193,25 67 0 79 16 0 0 0 0 198 0 0 0 4 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 16:07:09.006393,25 67 0 80 16 0 0 0 0 198 0 0 0 4 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 16:12:05.931162,25 67 0 80 16 0 0 0 0 198 0 0 0 5 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 17:01:12.468824,25 67 0 80 16 0 0 0 0 198 0 0 0 5 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 17:07:07.219155,25 67 0 81 16 0 0 0 0 198 0 0 0 5 47 0 0 0 0 0 0 0 0 0 0
2025-02-27 17:12:05.303046,25 67 0 81 16 0 0 0 0 198 0 0 0 6 47 0 0 0 0 0 0 0 0 0 0

 (09-02-2025 )67 25 64 0 251 15 0 0 0 0 195 0 0 0 176 46 0 0 0 0 0 0 0 0 0 0 240
 (09-02-2025) 67 25 65 0 1 16 0 0 0 0 196 0 0 0 182 46 0 0 0 0 0 0 0 0 0 0 21

Byte 3 and 4 is an hour meter : 16*256+81 = 4177 hours but i don't know what that corresponds to.

tspopp commented:
I added those, good find :+1: :smile:

FSlunkk commented:
2025-02-27 15:49:37.102192,31 218 0 237 0 7 2 68 1 139 128 89 0 14 1 200 0 0 1 0 0 2 0 0 0 0 0 0 0 0 0
Byte n°1 is Electric current intensity : 218 --> 2,18 A
Byte n°9, 10, 11 (and 12?) Total Energy Counter : 139 + 128*256 + 89*156² = 5 865 611 Wh
I will continue my research tomorrow.

FSlunkk commented:
I think there is a problem with time.
![image](https://github.com/user-attachments/assets/009765eb-303a-4156-89bd-0c7f979a9188)
2025-02-27 09:28:59.570912,29 54 2 0 33 0 0 0 0 0 0 0 0 0 0 0 0 59 50 **31 9** 0 0 0 0 65 66 0 0
Byte 19 is minutes
Byte 20 is hour.

FSlunkk commented:
I think that i can't find more things. Is there something important that i missed?

tspopp commented:
I added your latest findings and fixed the time issue. I will proceed with enabling MITM for you.

A few thing are important to add (if you have them):

- Enabling "Additional electric backup aka heat-element". There should be an switch in the engineering menu, to deactivate the heatpump: ![signal-2025-03-03-200033_005](https://github.com/user-attachments/assets/385e55fb-0dfb-46d7-933d-1aaa61fcc44c) We need this for the PV Modes "HeatPump only". It should be in the HMI Message (194).

- States in Main Message: If you turn on the heatpump, first the fans will turn on, then there will be some "defrost" for a few seconds and then the heatpump will start. This is represented in several icons shown in the HMI. A little fan appears, or a symbol for the heat-element. These states are within the Main Status Message (193). You should be able to identify those when you start the heatpump manually, e.g by using BOOST:
![signal-2025-03-03-200033_003](https://github.com/user-attachments/assets/170a8944-191d-46d9-9163-20d810e758e8)

![signal-2025-03-03-200033_002](https://github.com/user-attachments/assets/9aca5ff8-adb2-437e-9389-558f01781eba)

FSlunkk commented:
There is not "Additional electric backup" in my odyssee.
There is not icons on the hmi but i find that looking at the sensor information :
Byte 17 : 
1 = Heating element (observed via testmode)
24 = fan on
26 = fan + compressor = Heatpump
27 = heatpump + heating element
32 = defrost
56 = fan + defrost

FSlunkk commented:
For byte 17 in 193 : 
58 : heatpump + defrost

disable heatelement
I don't have the possibility to disable the use of the heat element.

states: pv-input, solar-input
There is not these parameters, my heatpump is surely too old.

timer window a/b
I set first 10:00 to 14:00 and from 16:00 to 22:00 but there is nothing that seems to appear.
2025-03-04 22:16:56.969438,29 54 18 0 33 0 0 0 0 0 0 0 0 0 0 0 0 68 50 19 22 0 0 0 0 65 66 0 0
2025-03-04 22:17:46.363428,29 54 18 0 33 0 0 0 0 0 0 0 0 0 0 0 0 68 50 20 22 0 0 0 0 65 66 0 0
2025-03-04 22:18:46.151666,29 54 18 0 33 0 0 0 0 0 0 0 0 0 0 0 0 68 50 21 22 0 0 0 0 65 66 0 0
And i set after 11:00 to 16:00 and 20:00 to 03.00 and there is a countdown in minutes that corresponds to the end of the range.
2025-03-04 22:19:41.972210,29 54 18 4 33 0 0 0 0 23 1 0 0 0 0 0 0 68 50 21 22 0 0 0 0 65 66 0 0
2025-03-04 22:19:46.231119,29 54 18 4 33 0 0 0 0 22 1 0 0 0 0 0 0 68 50 22 22 0 0 0 0 65 66 0 0
2025-03-04 22:20:46.383805,29 54 18 4 33 0 0 0 0 21 1 0 0 0 0 0 0 68 50 23 22 0 0 0 0 65 66 0 0
2025-03-04 22:21:46.113033,29 54 18 4 33 0 0 0 0 20 1 0 0 0 0 0 0 68 50 24 22 0 0 0 0 65 66 0 0
2025-03-04 22:22:46.104009,29 54 18 4 33 0 0 0 0 19 1 0 0 0 0 0 0 68 50 25 22 0 0 0 0 65 66 0 0
2025-03-04 22:23:46.098194,29 54 18 4 33 0 0 0 0 18 1 0 0 0 0 0 0 68 50 26 22 0 0 0 0 65 66 0 0
2025-03-04 22:24:46.097257,29 54 18 4 33 0 0 0 0 17 1 0 0 0 0 0 0 68 50 27 22 0 0 0 0 65 66 0 0
2025-03-04 22:25:46.104204,29 54 18 4 33 0 0 0 0 16 1 0 0 0 0 0 0 68 50 28 22 0 0 0 0 65 66 0 0
2025-03-04 22:26:46.065974,29 54 18 4 33 0 0 0 0 15 1 0 0 0 0 0 0 68 50 29 22 0 0 0 0 65 66 0 0
2025-03-04 22:27:46.073615,29 54 18 4 33 0 0 0 0 14 1 0 0 0 0 0 0 68 50 30 22 0 0 0 0 65 66 0 0
2025-03-04 22:28:46.225120,29 54 18 4 33 0 0 0 0 13 1 0 0 0 0 0 0 68 50 31 22 0 0 0 0 65 66 0 0
2025-03-04 22:29:46.056319,29 54 18 4 33 0 0 0 0 12 1 0 0 0 0 0 0 68 50 32 22 0 0 0 0 65 66 0 0
2025-03-04 22:30:46.044842,29 54 18 4 33 0 0 0 0 11 1 0 0 0 0 0 0 68 50 33 22 0 0 0 0 65 66 0 0
2025-03-04 22:31:46.022869,29 54 18 4 33 0 0 0 0 10 1 0 0 0 0 0 0 68 50 34 22 0 0 0 0 65 66 0 0
2025-03-04 22:32:46.028158,29 54 18 4 33 0 0 0 0 9 1 0 0 0 0 0 0 68 50 35 22 0 0 0 0 65 66 0 0
2025-03-04 22:33:46.157770,29 54 18 4 33 0 0 0 0 8 1 0 0 0 0 0 0 68 50 36 22 0 0 0 0 65 66 0 0
2025-03-04 22:34:45.996328,29 54 18 4 33 0 0 0 0 7 1 0 0 0 0 0 0 68 50 37 22 0 0 0 0 65 66 0 0
2025-03-04 22:35:45.986927,29 54 18 4 33 0 0 0 0 6 1 0 0 0 0 0 0 68 50 38 22 0 0 0 0 65 66 0 0
2025-03-04 22:36:45.987611,29 54 18 4 33 0 0 0 0 5 1 0 0 0 0 0 0 68 50 39 22 0 0 0 0 65 66 0 0
2025-03-04 22:37:45.970696,29 54 18 4 33 0 0 0 0 4 1 0 0 0 0 0 0 68 50 40 22 0 0 0 0 65 66 0 0
2025-03-04 22:38:45.960126,29 54 18 4 33 0 0 0 0 3 1 0 0 0 0 0 0 68 50 41 22 0 0 0 0 65 66 0 0
2025-03-04 22:39:45.947918,29 54 18 4 33 0 0 0 0 2 1 0 0 0 0 0 0 68 50 42 22 0 0 0 0 65 66 0 0
2025-03-04 22:40:45.937913,29 54 18 4 33 0 0 0 0 1 1 0 0 0 0 0 0 68 50 43 22 0 0 0 0 65 66 0 0
2025-03-04 22:41:45.927372,29 54 18 4 33 0 0 0 0 0 1 0 0 0 0 0 0 68 50 44 22 0 0 0 0 65 66 0 0
2025-03-04 22:42:45.915372,29 54 18 4 33 0 0 0 0 255 0 0 0 0 0 0 0 68 50 45 22 0 0 0 0 65 66 0 0
2025-03-04 22:43:45.902424,29 54 18 4 33 0 0 0 0 254 0 0 0 0 0 0 0 68 50 46 22 0 0 0 0 65 66 0 0
2025-03-04 22:44:45.882334,29 54 18 4 33 0 0 0 0 253 0 0 0 0 0 0 0 68 50 47 22 0 0 0 0 65 66 0 0
2025-03-04 22:45:45.849225,29 54 18 4 33 0 0 0 0 252 0 0 0 0 0 0 0 68 50 48 22 0 0 0 0 65 66 0 0


fan exhaust config (INT/EXT)
If i set your machine to AirDuct mode INT/EXT, i don't have an additional menu.

tspopp commented:
Hey @FSlunkk , finally I've been able to test the rewritten MITM code, with my legacy heatpump it works as expected. With my latest update you may try to:

- put your board into MITM configuration (no jumper set)
- flash latest build from this branch, configuration is already set to MITM
- do the following tests at the hmi controller:
  - [ ] verify that temperature values are still shown in the HMI
  - [ ] verify that setting the heatpump to BOOST will actually turn the heatpump / heatelement on
  - [ ] verify that the list of errors is still available when opening the super secret menu

tspopp commented:
![options](https://github.com/user-attachments/assets/989ca71b-3294-4e61-9e4b-e9f4f60c7afc)

Do you have these menu items, too? In PV Mode this installation option is overriden to "heat pump only", so it won't use any external connected boilers. I am not sure, if this is even possible with your heatpump :shrug: 

If you have it, we should find it in the hmi message 194.

FSlunkk commented:
I don't have these menu items.

FSlunkk commented:
I tested it, everything works fine.

tspopp commented:
🎉

tspopp commented:
Hey @FSlunkk, please test the latest version, it should be feature complete.

You may want to test if overrides in MITM mode are working as expected, either from HomeAssistant (if you use that) or directly by publishing MQTT values to the [MQTT control topics](https://github.com/tspopp/AquaMQTT/blob/main/MQTT.md).

Moreover I added a fix for the [PV Mode Heat Pump](https://github.com/tspopp/AquaMQTT/blob/main/README-PV.md). If you set the flag `flagPVModeHeatPump` via MQTT to true, your heatpump should heat up to maximum temperature without using the heating element. The behaviour might be a little different, since your heatpump does not offer the possibility to explicit disable the heating element. In your case I just set the heatpump to "MAN ECO ON" (eco on should not use heat element) and set the target temperature to 62°C / maximum configured temperature. Hope this will do the trick :)

FSlunkk commented:
Hey,
I tested MITM mode from Home Assistant and it doesn't work.

tspopp commented:
So, you select for example operation mode BOOST and nothing changes? :thinking:  Do you see messages published to the control topics in MQTT explorer? Do you see any overrides being set to true? Do you see that the operation mode in the MQTT hmi section actualy changes to BOOST?

tspopp commented:
> But nothing changed on the hmi.
Yeah, AquaMQTT cannot change what is shown on the HMI. AquaMQTT is overriding values sent from the HMI Controller. Thus, the HMI Controller does not know if we are overriding the operation mode or the target temperature.

The imporant point is, if your heat pump is starting to heat. My heatpump actually makes a beep sound when I enable BOOST mode.

> The Water temp target changed.
This is good, in BOOST mode the water target temp is automatically overriden to 62°C, which is the max temperature which can be configured in the Configuration.h file.

> But when i tried to set to 54°C, there is no change.
This is expected in BOOST mode, since it is fixed to 62°C. If you go back to "Man Eco On" it should be possible to set the water temperature.

FSlunkk commented:
In boost mode, the heatpump doesn't start.
I tried the other modes, i see the modes appear in ctrl but not "Manual Eco off"
![image](https://github.com/user-attachments/assets/46a64874-411b-4635-8f53-763d7a61798a)

tspopp commented:
Hey @FSlunkk, to debug this further I need more information. Posting screenshots of MQTT Explorer is ok, but can you change the  configuration value to

```
constexpr uint32_t MQTT_FULL_UPDATE_MS = 1000;
```

then all values will be updated frequently and we see what is going on after posting something to the ctrl topic.

If I understood correctly, it is not possible to change the operationMode shown in the hmi section? Is this correct?

tspopp commented:
@FSlunkk: I think I will merge this PR soon, since it contains a few changes important for the new hardware revision. But don't worry, I still want to continue debugging the last issues you have. Since MITM is already working, I guess it will be easy to fix the leftover issues.
